### PR TITLE
refactor: Upload use batch start

### PR DESCRIPTION
**Pull request by zombieJ**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [x] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |    Upload support `beforeUpload` return `Upload.LIST_IGNORE` to skip item added in UploadList.    |
| 🇨🇳 Chinese |   Upload 支持 `beforeUpload` 返回 `Upload.LIST_IGNORE` 时不将文件展示在列表中。       |

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

File: components/upload/Upload.tsx [hunk 1]
```diff
@@ -92,19 +92,50 @@ const InternalUpload: React.ForwardRefRenderFunction<unknown, UploadProps> = (pr
...
+    });
+
+    onInternalChange({
+      // [Legacy] This file is filled by `beforeUpload` before.
+      // We move start status in `onBatchStart` instead which means we only trigger one file.
+      // It's strange since the first trigger of file is File instance
+      // but next will be a wrapped object.
+      // Seems always return wrapped object is more make sense.
+      // https://github.com/ant-design/ant-design/issues/10293
+      file: filteredFileInfoList[0]?.file,
```
Comment (zombieJ) [inline components/upload/Upload.tsx:129]: 这里挺奇怪，其他地方都是封装过的对象，只有它是不封装的：
https://github.com/ant-design/ant-design/issues/10293

File: components/upload/Upload.tsx [hunk 2]
```diff
@@ -203,41 +235,46 @@ const InternalUpload: React.ForwardRefRenderFunction<unknown, UploadProps> = (pr
...
-      // Get unique file list
-      const uniqueList: UploadFile<any>[] = [];
-      getFileList()
-        .concat(fileListArgs.map(fileToObject))
-        .forEach(f => {
-          if (uniqueList.every(uf => uf.uid !== f.uid)) {
-            uniqueList.push(f);
-          }
+  const mergedBeforeUpload = async (file: RcFile, fileListArgs: RcFile[]) => {
+    const { beforeUpload, transformFile } = props;
```
Comment (kerm1it) [inline components/upload/Upload.tsx:239]: rc 里面已经删除了 `transformFile`，antd 里面就也废弃掉吧，只保持兼容，推荐用 `beforeUpload` 替换。

Comment (zombieJ) [reply components/upload/Upload.tsx:225]: 嗯，这个 `transformFile` 是在 `beforeUpload` 里模拟的哈~~

Comment (kerm1it) [reply components/upload/Upload.tsx:225]: 刚看文档已经废弃了，我以为还没废弃呢～

**Comment by zombieJ** [comment]
诡异，这么随机的错误……

**Review by I-take-it-smart — APPROVED** [review]

**Review by xrkffgg — APPROVED** [review]

File: package.json [hunk 3]
```diff
@@ -147,7 +147,7 @@
     "rc-tree": "~4.1.0",
     "rc-tree-select": "~4.3.0",
     "rc-trigger": "^5.2.1",
-    "rc-upload": "~3.3.4",
+    "rc-upload": "~4.0.0-alpha.6",
```
Comment (kerm1it) [inline package.json:150]: 现在先不发 4.0.0 正式版本么？

**Comment by alienzhangyw** [comment]
beforeUpload的typescript类型需要更新？
``` typescript
beforeUpload={(file) => {
        if (file.type !== 'image/png') {
          message.error(`${file.name} is not a png file`);
        }
        return file.type === 'image/png' ? true : Upload.LIST_IGNORE;
      }}
```
Eslint：
```
不能将类型“(file: RcFile) => {}”分配给类型“(file: RcFile, FileList: RcFile[]) => boolean | Promise<void | File | Blob>”。
  不能将类型“{}”分配给类型“boolean | Promise<void | File | Blob>”。
    类型“{}”缺少类型“Promise<void | File | Blob>”中的以下属性: then, catch, [Symbol.toStringTag], finally
```

**Pull request merged**